File: 2023/day-02.rb
require 'set'
def parse_input(input)
    games = {}

    File.foreach(input) do |line|
        one = line.chomp.split(":")
        game_id = one[0].split(" ").last.to_i
        
        matches = one[1].strip.split(";")
  
        match_array = []
        matches.each_with_index do |match, idx|
            
            cubes = match.strip.split(",")
            
            cubes_array = []
            cubes.each do |cube|
                num, color = cube.strip.split(" ")
                cubes_array.push([color,num.to_i])
            end
            #p cubes_array
            match_array.push(cubes_array)
        end
        #p match_array
        games[game_id]=(match_array)
        
    end
    games
end

def calc_possible_games(games)
  
    actual_cubes = {}
    actual_cubes["red"] = 12
    actual_cubes["green"] = 13
    actual_cubes["blue"] = 14
    total = 0
    #puts "** p games **"
    #p games

    #puts "** p games[0] **"
    #p games[1]

   # p all_games
    for i in 1..100 do
        game = games[i]
        red, green, blue = 0, 0, 0
        possible = true
        
        game.each do |match|
            match.each do |reveal|
                if reveal[0] == "red"
                    red = reveal[1] unless red > reveal[1]
                elsif reveal[0] == "green"
                    green = reveal[1] unless green > reveal[1]
                elsif reveal[0] == "blue"
                    blue = reveal[1] unless blue > reveal[1]
                end
            end
        end
        total += red * blue * green
        puts "red x blue x green: #{red} * #{blue} * #{green} = #{red * blue * green}"
    end
    puts total
end

input = "Inputs/day-02.txt"
games = {}
games = parse_input(input)

possible_game_id_total = calc_possible_games(games)

File: 2023/day_02.py
def read_input(path):
    with open(path) as f:
        return f.read()
    
def parse_file(raw):
    games = {}
    for line in raw.splitlines():
        temp = line.split(": ")

        _game_id = temp[0].split(" ")
        game_id = int(_game_id[1])

        _matches = temp[1].split("; ")
        match_array = []
        for _match in _matches:
            _play = _match.split(", ")
            play_array = []

            for _cubes in _play:
                num, color = _cubes.split(" ")
                play_array.append([int(num), color])
            match_array.append(play_array)
        games[game_id] = match_array
    return games

def solve_part1(games):
    game_data = {"red": 12, "green": 13, "blue": 14}
    total = 0

    for key, value in games.items():
        used = {"red": 0, "green": 0, "blue": 0}
        valid = True
        for round in value:
            for play in round:
                used[play[1]] += play[0]
                if game_data[play[1]] < play[0]:
                    valid = False
                    break
        if valid:
            total += key

    return total

def solve_part2(games):
    total = 0

    for key, value in games.items():
        used = {"red": None, "green": None, "blue": None}
        for round in value:
            for play in round:
                if used[play[1]] == None or used[play[1]] < play[0]:
                    used[play[1]] = play[0]

        total += used["red"] * used["blue"] * used["green"]

    return total


path = "Inputs/day-02.txt"
raw = read_input(path)
data = parse_file(raw)
print("part 1: " + str(solve_part1(data)))
print("part 2: " + str(solve_part2(data)))